common/libs/framing/frame_detector.h

```diff
@@ -0,0 +1,129 @@
+#ifndef __FRAME_DETECTOR__
+#define __FRAME_DETECTOR__
+
+#include "serial_listeners.h"
+#include <stdint.h>
+#include <string.h>
+
+template <class RxBuffer, int FRAME_BUF_LEN>
+class FrameDetector : public RxListener {
```

Comment:
Please add some high-level comments as to how this works, e.g. a couple of typical situations explained in terms of what state transitions they cause.

Reply:
While writing these I have realized there might be some smelly stuff going on when the full frame is received - we will have a race condition between `OnCharacterMatch` and `OnRxComplete` callbacks. I will test against that and fix it if something comes up.

Reply:
Ok, we are good in the all good case. I was wrong assuming that both `OnCharacterMatch` and `OnRxComplete` will come at the same time. We will always have one char place extra in the buffer when the max size frame comes, because we discard the frame start marker (later we also discard end frame marker in ProcessReceivedData()).

Though, in case sender sends a frame that's 1 byte longer than max frame length, we *will* have this race condition. If  `OnCharacterMatch` wins we will have frame detected and end up in LOST state. If `OnRxComplete` wins we will have no frame detected and end up in `WAIT_FOR_START_MARKER` state. In both cases we are able to recover, besides sender was already wrong in sending a frame too long. We can resolve this race condition by setting hardware interrupt priorities. And in soft implementation of RxBuffer we can resolve this by explicitly coding the order in which callbacks are called like it is done now in `SoftRxBuffer`.

Reply:
After putting more thinking to it, I think I should not be discarding the start marker, so that the whole thing is not driven by side effects. I am afraid that will have implications on code clarity, but still...

Reply:
I have refactored the FrameDetector to be more explicit in handling markers. Now we don't have the bug with illegaly long frame being detected. Also I have added a testing infrastructure for fuzzing as you have requested.